### PR TITLE
8266086: [lworld][lw3] C1 produces incorrect code when GlobalValueNumbering is used

### PR DESCRIPTION
**Pull request by fparain**
Please review this fix in C1 GlobalValueNumbering.
The problem is that C1 doesn't track that a flattened field has been updated when it writes the individual values of this flattened field. The proposed fix is to record the enclosing flattened field with the StoreField node of each individual field, and use this information when the GlobalValueNumbering processes those nodes to kill the flattened field in the ValueMap.

Tested locally (Linux 64) with hotspot_valhalla (including new unit test) and jdk_valhalla test suites.

Thank you,

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8266086](https://bugs.openjdk.java.net/browse/JDK-8266086): [lworld][lw3] C1 produces incorrect code when GlobalValueNumbering is used

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/395/head:pull/395` \
`$ git checkout pull/395`

Update a local copy of the PR: \
`$ git checkout pull/395` \
`$ git pull https://git.openjdk.java.net/valhalla pull/395/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 395`

View PR using the GUI difftool: \
`$ git pr show -t 395`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/395.diff">https://git.openjdk.java.net/valhalla/pull/395.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java:2]: Copyright should be 2021.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+package compiler.valhalla.inlinetypes;
+
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @summary Test value numbering behaves correctly with flattened fields
+ * @library /testlibrary /test/lib /compiler/whitebox /
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java:32]: Whitebox is not used.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * @compile TestC1ValueNumbering.java
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java:33]: Not required.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * @run main/othervm -Xcomp -XX:TieredStopAtLevel=1 -ea -XX:+UseGlobalValueNumbering
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java:34]: `UseGlobalValueNumbering` is true by default.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ *                   -XX:CompileCommand=compileonly,compiler.valhalla.inlinetypes.TestC1ValueNumbering::*
+ *                   compiler.valhalla.inlinetypes.TestC1ValueNumbering
+ */
+
+public class TestC1ValueNumbering {
+  static primitive class Point {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1ValueNumbering.java:40]: 4x whitespace indentation should be used for Java code.

File: src/hotspot/share/c1/c1_ValueMap.hpp [hunk 6]
```diff
@@ -148,6 +148,9 @@ class ValueNumberingVisitor: public InstructionVisitor {
       kill_memory();
     } else {
       kill_field(x->field(), x->needs_patching());
+      if (x->enclosing_field() != NULL) {
+        kill_field(x->enclosing_field(), true);
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_ValueMap.hpp:152]: Why do you need `all_offsets == true`? The offset of the enclosing field should always be known, right?

**Review by TobiHartmann — COMMENTED** [review]
Good catch! I've added some comments.

**Comment by fparain** [comment]
Hi Tobias,

Thank you for the review.
Most of the issues you spotted have been fixed on the latest commit.
Regarding the `all_offset == true` issue, it is needed to invalidate all field entries in the ValueMap. We would like to only invalidate the entries of the nested fields related to the flattened field. But there's no infrastructure yet to easily identity which fields are related to a given flattened field. So the conservative approach is to invalidate all field entries.

Regards,

Fred

**Review by TobiHartmann — APPROVED** [review]
Thanks for updating, looks good to me!

Best regards,
Tobias

**Pull request closed without merging**